xbmc/filesystem/NFSFile.cpp

```diff
@@ -585,34 +599,34 @@ bool CNFSFile::Open(const CURL& url)
 
   std::unique_lock<CCriticalSection> lock(gNfsConnection);
 
-  auto NfsOpen = [this](const CURL& url, std::string& filename) -> bool
-  {
-    if (!gNfsConnection.Connect(url, filename))
-      return false;
+  if (!gNfsConnection.Connect(url, filename))
+    return false;
 
-    m_pNfsContext = gNfsConnection.GetNfsContext();
-    m_exportPath = gNfsConnection.GetContextMapId();
+  m_pNfsContext = gNfsConnection.GetNfsContext();
+  m_exportPath = gNfsConnection.GetContextMapId();
 
-    return nfs_open(m_pNfsContext, filename.c_str(), O_RDONLY, &m_pFileHandle) == 0;
-  };
+  int ret = nfs_open(m_pNfsContext, filename.c_str(), O_RDONLY, &m_pFileHandle);
 
-  if (!NfsOpen(url, filename))
+  if (ret == -11) // NFS4ERR_EXPIRED
   {
     CLog::Log(LOGERROR,
               "CNFSFile::Open: Unable to open file - trying again with a new context: error: '{}'",
               nfs_get_error(m_pNfsContext));
 
     gNfsConnection.Deinit();
+    m_pNfsContext = gNfsConnection.GetNfsContext();
+    m_exportPath = gNfsConnection.GetContextMapId();
+    ret = nfs_open(m_pNfsContext, filename.c_str(), O_RDONLY, &m_pFileHandle);
+  }
 
-    if (!NfsOpen(url, filename))
-    {
-      CLog::Log(LOGERROR, "CNFSFile::Open: Unable to open file: '{}' error: '{}'",
-                url.GetFileName(), nfs_get_error(m_pNfsContext));
+  if (ret != 0)
```

Comment:
Are positive error codes allowed?

Reply:
only `ret == 0` is open successfully, then `ret != 0` is open FAILED which is precisely what we want to detect here